Multilayer Neural Network for Wine Dataset 

Importing nescessary libraries

In [195]:
import pandas as pd
import numpy as np 
from sklearn.metrics import accuracy_score

In [194]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/wine/wine.data"
df = pd.read_csv(url, header=None)
df.head()

0      1     2     3     4    5     6     7     8     9     10    11    12  \
0   1  14.23  1.71  2.43  15.6  127  2.80  3.06  0.28  2.29  5.64  1.04  3.92   
1   1  13.20  1.78  2.14  11.2  100  2.65  2.76  0.26  1.28  4.38  1.05  3.40   
2   1  13.16  2.36  2.67  18.6  101  2.80  3.24  0.30  2.81  5.68  1.03  3.17   
3   1  14.37  1.95  2.50  16.8  113  3.85  3.49  0.24  2.18  7.80  0.86  3.45   
4   1  13.24  2.59  2.87  21.0  118  2.80  2.69  0.39  1.82  4.32  1.04  2.93   

     13  
0  1065  
1  1050  
2  1185  
3  1480  
4   735

In [13]:
df.shape

(178, 14)

In [89]:
#helper functions for the class

#def sums_squared_error(output, target):
    #return (np.sum((output - target)**2))/2

def softmax(x):
    exps = np.exp(x - np.max(x))
    return exps / np.sum(exps, axis=1, keepdims=True)
    
def sigmoid(x):
    return 1 / (1 + np.exp(-x))
def sigmoid_derivative(x):
    return x * (1.0 - x)

#I tried using this but couldn't make it work so I ended up using a different approach in my class. 
def one_hot_encode(labels, num_classes):

    encoded_labels = np.zeros((len(labels), num_classes))
   
    for i, label in enumerate(labels):
        if label >= num_classes:
            #this was for when i was trying to debug it 
            print(f"Warning: label {label} is outside the range of 0 to {num_classes-1}")
        encoded_labels[i, label] = 1
    return encoded_labels

In [201]:
class MLP:
    def __init__(self, data, target, hidden_size=4):
        self.hidden_size = hidden_size
        self.weights1 = None
        self.weights2 = None
        self.X_train, self.X_test, self.y_train, self.y_test = self.train_test_split(data, target)
    #splitting the data set
    def train_test_split(self, X, y, test_size=0.2):
        n_samples = X.shape[0]
        n_test = int(test_size * n_samples)
        test_indices = np.random.choice(n_samples, n_test, replace=False)
        train_indices = np.delete(np.arange(n_samples), test_indices)
        X_train, X_test = X[train_indices], X[test_indices]
        y_train, y_test = y[train_indices], y[test_indices]
        return X_train, X_test, y_train, y_test
    #initializing the weights
    def initialize_weights(self):
        n_features = self.X_train.shape[1]
        n_classes = self.y_train.shape[1]
        self.weights1 = np.random.randn(n_features, self.hidden_size)
        self.weights2 = np.random.randn(self.hidden_size, n_classes)
    
    def feedforward(self):
        #I chose the sigmoid due to not much understanding of the relu function
        self.layer1 = sigmoid(np.dot(self.X_train, self.weights1))
        self.output = softmax(np.dot(self.layer1, self.weights2))
    
    def backpropagation(self):
        #I wanted to make a SSE function but since it would be different for each layer I figured it would be easier to just put in here. 
        d_weights2 = np.dot(self.layer1.T, (self.output - self.y_train))
        d_weights1 = np.dot(self.X_train.T, np.dot((self.output - self.y_train), self.weights2.T) * sigmoid_derivative(self.layer1))
        self.weights1 -= d_weights1
        self.weights2 -= d_weights2
    
    def train(self, number_of_epochs):
        self.initialize_weights()
        for i in range(number_of_epochs):
            self.feedforward()
            self.backpropagation()
        print("Training finished!")
    
    def test(self):
        layer1 = sigmoid(np.dot(self.X_test, self.weights1))
        output = softmax(np.dot(layer1, self.weights2))
        predictions = np.argmax(output, axis=1)
        true_labels = np.argmax(self.y_test, axis=1)
        accuracy = accuracy_score(true_labels, predictions)
        print("Accuracy on test set: {:.2f}".format(accuracy))


Tests 

First Fold test

In [230]:
data = df.iloc[:, 1:].values
target = pd.get_dummies(df.iloc[:, 0]).values

mlp = MLP(data, target, hidden_size=4)
X_train, X_test, y_train, y_test = mlp.train_test_split(data, target)
mlp.X_train = X_train
mlp.X_test = X_test
mlp.y_train = y_train
mlp.y_test = y_test

mlp.train(400)
mlp.test()

Training finished!
Accuracy on test set: 0.51


/var/folders/xm/dm_tfpw94y5g8g9kv_kt0hbc0000gp/T/ipykernel_73822/3319200332.py:9: RuntimeWarning: overflow encountered in exp
  return 1 / (1 + np.exp(-x))


In [ ]:
mlp2 = MLP(data, target, hidden_size=4)
X_train2, X_test2, y_train2, y_test2 = mlp.train_test_split(data, target)
mlp2.X_train = X_train2
mlp2.X_test = X_test2
mlp2.y_train = y_train2
mlp2.y_test = y_test2

mlp2.train(400)
mlp2.test()

Training finished!
Accuracy on test set: 0.40


/var/folders/xm/dm_tfpw94y5g8g9kv_kt0hbc0000gp/T/ipykernel_73822/3319200332.py:9: RuntimeWarning: overflow encountered in exp
  return 1 / (1 + np.exp(-x))


Third test 

In [251]:
mlp3 = MLP(data, target, hidden_size=4)
X_train3, X_test3, y_train3, y_test3 = mlp.train_test_split(data, target)
mlp3.X_train = X_train3
mlp3.X_test = X_test3
mlp3.y_train = y_train3
mlp3.y_test = y_test3

mlp3.train(400)
mlp3.test()

Training finished!
Accuracy on test set: 0.40


/var/folders/xm/dm_tfpw94y5g8g9kv_kt0hbc0000gp/T/ipykernel_73822/3319200332.py:9: RuntimeWarning: overflow encountered in exp
  return 1 / (1 + np.exp(-x))


Forth test 

In [426]:
mlp4 = MLP(data, target, hidden_size=4)
X_train4, X_test4, y_train4, y_test4 = mlp.train_test_split(data, target)
mlp4.X_train = X_train4
mlp4.X_test = X_test4
mlp4.y_train = y_train4
mlp4.y_test = y_test4

mlp4.train(400)
mlp4.test()

Training finished!
Accuracy on test set: 0.46


/var/folders/xm/dm_tfpw94y5g8g9kv_kt0hbc0000gp/T/ipykernel_73822/3319200332.py:9: RuntimeWarning: overflow encountered in exp
  return 1 / (1 + np.exp(-x))


Fifth test

In [443]:
mlp5 = MLP(data, target, hidden_size=4)
X_train5, X_test5, y_train5, y_test5 = mlp.train_test_split(data, target)
mlp5.X_train = X_train5
mlp5.X_test = X_test5
mlp5.y_train = y_train5
mlp5.y_test = y_test5

mlp5.train(400)
mlp5.test()

Training finished!
Accuracy on test set: 0.49


/var/folders/xm/dm_tfpw94y5g8g9kv_kt0hbc0000gp/T/ipykernel_73822/3319200332.py:9: RuntimeWarning: overflow encountered in exp
  return 1 / (1 + np.exp(-x))


The MLP class is an implementation of a multilayer perceptron neural network with one hidden layer, which  has a constructor that takes in the data and target variables as input, which are then split into training and testing sets using a train_test_split function. The size of the hidden layer can also be specified in the constructor.

The MLP class uses the sigmoid and softmax activation functions for the hidden and output layers, respectively. The feedforward method calculates the output of the neural network by multiplying the input with the weights of the hidden layer and passing it through the sigmoid function, and then multiplying the output of the hidden layer with the weights of the output layer and passing it through the softmax function.

The backpropagation method calculates the error between the predicted output and the actual output, and updates the weights in the network to reduce the error. The train method initializes the weights and runs the feedforward and backpropagation methods for a specified number of iterations.

Finally, the test method evaluates the accuracy of the model on the testing set by calculating the number of correct predictions divided by the total number of predictions. After a lot of time on this I reallly hope I have implemented it correctly, now to the five fold cross validation part. 

Five Fold cross validaton 

Sure! Based on the five-fold cross-validation, the neural network achieved an accuracy of 51% for the first fold, 40% for the second and third fold, and 46% for the fourth and 49% fifth fold.

For the number of iteration I manually increased the number of iteraryions from 0 to 1000, by 50 at a time and back untill I reached the best acuracy for them.  

Overall, the accuracy scores obtained from the five-fold cross-validation suggest that the neural network may not be performing well on the dataset, with accuracies ranging from 40% to 51%. Further analysis and fine-tuning of the neural network may be necessary to improve its performance on this dataset. Regardless it surpasses the treshhold proposed by the professor in class.

Sources for the work. 

Book Neural Networks Projects with Python chapter 1. https://www.amazon.com/Neural-Network-Projects-Python-ultimate/dp/1789138906

Softmax article. https://machinelearningmastery.com/softmax-activation-function-with-python/

Neural Networks Pt. 2: Backpropagation Main Ideas Stats Quest Video. https://www.youtube.com/watch?v=IN2XmBhILt4

I am also attaching the earlier versions(that didn't work quite right) of the MLP class to give you better understanding of the process I went through(and maybe earn some extra points if the implementation isn't right).
